data/src/main/java/com/microsoft/azure/kusto/data/auth/endpoints/KustoTrustedEndpoints.java

```diff
@@ -0,0 +1,144 @@
+package com.microsoft.azure.kusto.data.auth.endpoints;
+
+import com.microsoft.azure.kusto.data.Ensure;
+import com.microsoft.azure.kusto.data.auth.CloudInfo;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import com.microsoft.azure.kusto.data.exceptions.KustoClientInvalidConnectionStringException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.*;
+import java.util.function.Predicate;
+
+/**
+ * A helper class to determine which DNS names are "well-known/trusted"'
+ * Kusto endpoints. Untrusted endpoints might require additional configuration
+ * before they can be used, for security reasons.
+ */
+public class KustoTrustedEndpoints {
+    private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    public static boolean enableWellKnownKustoEndpointsValidation = true;
+    private static final Map<String, FastSuffixMatcher> matchers;
+    private static FastSuffixMatcher additionalMatcher;
+    private static Predicate<String> overrideMatcher; // We could unify this with matchers, but separating makes debugging easier
+
+    static {
+        matchers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+
+        WellKnownKustoEndpointsData.getInstance().AllowedEndpointsByLogin.forEach((key, value) -> {
+            List<MatchRule> rules = new ArrayList<>();
+            value.AllowedKustoSuffixes.forEach(suffix -> rules.add(new MatchRule(suffix, false)));
+            value.AllowedKustoHostnames.forEach(hostname -> rules.add(new MatchRule(hostname, true)));
+            matchers.put(key, FastSuffixMatcher.Create(rules));
+        });
+
+        additionalMatcher = null;
+        overrideMatcher = null;
+    }
+
+    /**
+     * @param matcher Rules that determine if a hostname is a valid/trusted Kusto endpoint
+     *  (replaces existing rules). Null means "use the default policy".
+     */
+    public static void SetOverridePolicy(Predicate<String> matcher) {
```

Comment:
Shouldn't this be locked/synchronized too?

Reply:
no

Reply:
What if someone calls validateHostnameIsTrusted at the same time as this?

Reply:
then he will have either the new matcher or the old matcher.

Reply:
1.        if (overrideMatcher != null) {
2.           if (overrideMatcher.test(hostname)) {

1 passes, anthoer thread changes overrideMatcher to null, 2 runs, you have an NPE

Reply:
:(
fixed